File: 4. April/17. 988. Smallest String Starting From Leaf.cpp
// بِسْمِ اللهِ الرَّحْمٰنِ الرَّحِيْمِ
#include <bits/stdc++.h>
using namespace std;
#define all(x) x.begin(), x.end()
typedef vector<int> vi;
typedef vector<vi> vii;

string result = "";

void solve(TreeNode *root, string curr)
{
    if (root == NULL)
    {
        return;
    }

    curr = char(root->val + 'a') + curr; // This will take O(length of curr) but I have ignored this. You can condiser this as well in calculating T.C

    if (root->left == NULL && root->right == NULL)
    {
        if (result == "" || curr < result)
        {
            result = curr;
        }
        return;
    }

    solve(root->left, curr);
    solve(root->right, curr);
}

string smallestFromLeaf(TreeNode *root)
{
    solve(root, "");

    return result;
}

File: 4. April/17. 988. Smallest String Starting From Leaf.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution(object):
    result = ""
    def solve(self, root, curr):
        if(root == None):
            return
        
        curr = chr((root.val) + ord('a')) + curr
        
        if(root.left == None and root.right == None):
            if self.result == "" or curr < self.result:
                # update result
                self.result = curr
            
            return
        
        self.solve(root.left, curr)
        self.solve(root.right, curr)
    
    def smallestFromLeaf(self, root):
        """
        :type root: TreeNode
        :rtype: str
        """
        
        self.solve(root, "")
        
        return self.result